src/librustc/hir/lowering.rs

```diff
@@ -72,23 +72,49 @@ use syntax_pos::Span;
 
 const HIR_ID_COUNTER_LOCKED: u32 = 0xFFFFFFFF;
 
-pub struct LoweringContext<'a> {
+macro_rules! p {
+    ($this:expr, $e:expr) => ({
+        P::alloc($this.arena, $e)
```

Comment:
I wonder if this could be just `$this.alloc($e)` and maybe then we wouldn't need a macro?
Then again, I think the macro allows nested allocations (e.g. `p!(self, Foo(p!(self, Bar(...))))`) without borrow conflicts, since `$this.arena` can just be copied.